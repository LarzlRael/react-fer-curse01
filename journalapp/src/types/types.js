export const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] remove Error',

    uiStartLoading: '[UI] Start loading',
    uiFinishLoading: '[UI] Finish loading',


    notesAddNew: '[Notes] New note',
    notesActive: '[Notes] Set action note',
    notesLoad: '[Notes] Load notes',
    notesUpdated: '[Notes] Update note saved',
    notesFileUrl: '[Notes] Update image url',
    notesDelete: '[Notes] Update delete note',
    notesLogoutCleaning: '[Notes] Logout cleaning',

}