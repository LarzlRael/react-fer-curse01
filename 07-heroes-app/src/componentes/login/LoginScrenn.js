import React from 'react'

export const LoginScrenn = ({ history }) => {

    const handleLogin = () => {
        // create a push in the history routes
        // history.push('/')
        // create a reaplce in the history routes
        history.replace('/')
    }

    return (
        <div className="container mt-5">
            <h1>Login</h1>
            <hr />
            <h1>
                <button
                    onClick={handleLogin}
                    type="button" className="btn btn-primary">Login</button>
            </h1>
        </div>
    )
}
