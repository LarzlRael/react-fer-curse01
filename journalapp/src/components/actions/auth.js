import { types } from "../../types/types"
import { facebookAuthProvider, firebase, googleAuthProvider } from '../../firebase/firebase-Config'
import { finishLoading, startLoading } from "./ui"
import Swal from 'sweetalert2';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch(startLoading());

        firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {

                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());

            })
            .catch(e => {
                
                dispatch(finishLoading());
                Swal.fire('Error',e.message,'error');
            })

    }
}

export const startGoogleLogin = () => {

    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                dispatch(
                    login(user.uid, user.displayName)
                );
            })
    }
}
export const startFacebookLogin = () => {

    return (dispatch) => {
        firebase.auth().signInWithPopup(facebookAuthProvider)
            .then(userCred => {
                console.log(userCred);
            })
            .catch((error) => {
                // Handle Errors here.
                var errorCode = error.code;
                var errorMessage = error.message;
                // The email of the user's account used.
                var email = error.email;
                // The firebase.auth.AuthCredential type that was used.
                var credential = error.credential;

                // ...
            });
    }
}
export const login = (uid, displayName) => ({

    type: types.login,
    payload: {
        uid,
        displayName
    }

})

export const startRegisterWithEmailPassword = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({ displayName: name });

                dispatch(user.uid, user.displayName)
            })
            .catch(e => {
                console.log(e);
                Swal.fire('Error',e.message,'error');
            })
    }
}

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout())
    }
}

export const logout = () => ({
    type: types.logout
})