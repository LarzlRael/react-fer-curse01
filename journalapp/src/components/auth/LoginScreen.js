import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { login, startFacebookLogin, startGoogleLogin, startLoginEmailPassword } from '../actions/auth'

export const LoginScreen = () => {

    // will give us the ability to dispatch from anywhere without needing the store
    const dispatch = useDispatch();

    // getting the state from store
    const { loading } = useSelector(state => state.ui);

    const [formValues, handleInputChange] = useForm({
        email: 'xdxd@gmail.com',
        password: '123456'
    });
    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();

        console.log(email, password);
        // using the useDistpach
        dispatch(startLoginEmailPassword(email, password));
    }

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    }

    const handleFacebookLogin = () => {
        dispatch(startFacebookLogin());

    }

    return (
        <>
            <form
                onSubmit={handleLogin}
                className="animate__animated animate__fadeIn animate__faster"
            >
                <h3 className="auth__title">
                    Login
                </h3>
                <input
                    autoComplete="off"
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    value={email}
                    onChange={handleInputChange}
                />
                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={handleInputChange}
                />
                <button
                    className="btn btn-primary btn-block"
                    disabled={loading}
                    type="submit">
                    Login
                </button>
                <hr />

                <div className="auth__social-networks">
                    <p>Login with your social networks</p>
                    <div
                        onClick={handleGoogleLogin}
                        className="google-btn"
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>

                    <div
                        className="google-btn mb-5"
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/thumb/5/51/Facebook_f_logo_%282019%29.svg/1365px-Facebook_f_logo_%282019%29.svg.png" alt="facebook button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with Facebook</b>
                        </p>
                    </div>
                </div>
                <Link
                    className="link"
                    to="/auth/register"
                >Create new Account</Link>
            </form>
        </>
    )
}
