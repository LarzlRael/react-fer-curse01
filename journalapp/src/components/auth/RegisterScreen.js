import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';

import validator from 'validator'
import { removeError, setError } from '../actions/ui';
import { startRegisterWithEmailPassword } from '../actions/auth';

const RegisterScreen = () => {

    const dispatch = useDispatch();
    //get the state 
    const { msgError } = useSelector(state => state.ui);


    const [formValues, handleInputChange] = useForm({
        name: 'larzdo',
        email: 'xdxd@gmail.com',
        password: '123456',
        password2: '123456'
    });

    const { name, email, password, password2 } = formValues;

    const handleSubmit = (e) => {
        e.preventDefault();

        if (isFormValid()) {
            dispatch(startRegisterWithEmailPassword(email, password, name));
        }

    }

    const isFormValid = () => {
        if (name.trim().length === 0) {

            dispatch(setError('Name is required'));
            return false;

        } else if (!validator.isEmail(email)) {

            dispatch(setError('Email is not valid'));
            return false;

        } else if (password !== password2 || password.length <= 5) {
            dispatch(setError('password should be at least 5 caracteres and match each it'));
            return false;
        }
        dispatch(removeError())
        return true;
    }


    return (
        <div >
            <h3 className="auth__title">
                Register
            </h3>
            <form onSubmit={handleSubmit}>

                {
                    msgError &&
                    <div className="auth__alert-error">
                        {msgError}
                    </div>
                }
                <input
                    autoComplete="off"
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    value={name}
                    onChange={handleInputChange}
                />
                <input
                    autoComplete="off"
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    value={email}
                    onChange={handleInputChange}
                />
                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={handleInputChange}
                />
                <input
                    type="password"
                    placeholder="Confirm password"
                    name="password2"
                    className="auth__input"
                    value={password2}
                    onChange={handleInputChange}
                />
                <button
                    className="btn btn-primary btn-block mb-5"
                    type="submit">
                    Register
                </button>
                <hr />

                <Link
                    className="link"
                    to="/auth/login"
                >
                    Already register?
                </Link>
            </form>
        </div>
    )
}

export default RegisterScreen
