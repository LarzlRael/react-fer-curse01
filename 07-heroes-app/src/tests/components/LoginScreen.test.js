import { mount } from "enzyme"
import { AuthContext } from "../../auth/AuthContext"
import { LoginScrenn } from "../../componentes/login/LoginScrenn"
import { types } from "../../types/types"

describe('Testing in <LoginScreen />', () => {

    const history = {
        replace: jest.fn()
    }

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: false
        }
    }

    const wrapper = mount(
        <AuthContext.Provider value={contextValue}>

            <LoginScrenn history={history} />

        </AuthContext.Provider>
    );

    test('should display correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should realize the dispatch and the navigation', () => {

        wrapper.find('button').prop('onClick')();

        expect(contextValue.dispatch).toHaveBeenCalledWith({
            type: types.login,
            payload: {
                name: 'Larz'
            }
        });

        expect(history.replace).toHaveBeenCalled();
    })

    test('should save the last path in the localstorage', () => {

        const handleClick = wrapper.find('button').prop('onClick');

        handleClick();

        expect(contextValue.dispatch).toHaveBeenCalledWith({
            type: types.login,
            payload: {
                name: 'Larz'
            }
        });

        expect(history.replace).toHaveBeenCalledWith('/');

        localStorage.setItem('lastPath', '/dc');
        handleClick();

        expect(history.replace).toHaveBeenCalledWith('/dc');
    })




})
