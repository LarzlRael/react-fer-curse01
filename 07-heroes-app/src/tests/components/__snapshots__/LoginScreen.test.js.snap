// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <LoginScreen /> should display correctly 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <h1>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="button"
    >
      Login
    </button>
  </h1>
</div>
`;
